### ISSUE TITLE
JDK 11.0.8 Runtime#availableProcessors() returns cpu_request / 1024 in OpenShift

### ISSUE THREAD

**Issue by beikov** [AdoptOpenJDK/openjdk-support#194]
**Platform:**
Linux/CentOS7
**Architecture:**
x86
Please describe the issue that you are having with our binaries and if possible how to recreate it.

When I set no or a low cpu request on an OpenShift pod I always get 1 returned by `Runtime#availableProcessors()`. It seems that OpenJDK 11.0.8 calculates the available processors by doing `max(1, cpu_request / 1024)` although it IMO should use something like `cpu_limit == null ? physical_cores : (cpu_limit / 1024)`.

I'm testing with the docker image jboss/base-jdk:11 and essentially doing the following to test it:

```
cat <<EOF >test.sh
#!/usr/lib/jvm/java-11/bin/java --source 11
public class App {
    public static void main(String ...args) {
        System.out.println(Runtime.getRuntime().availableProcessors());
    }
}
EOF
chmod +x test.sh
./test.sh
```

The JDK8 I used before had support for detecting the cgroups limits and properly detected the cpu_limit which in my case was 8192 i.e. 8 cores. Can I somehow force the old behavior?

**Comment by aahlenst** [comment]
@jerboaa Didn't you work in that area (cgroups) recently?

**Comment by jerboaa** [comment]
@beikov Yes. Intended behaviour. See this bug for the rationale:
https://bugs.openjdk.java.net/browse/JDK-8216366

**Comment by jerboaa** [comment]
By setting CPU requests in OpenShift you are basically setting a cgroups CPU limit, which OpenJDK respects.

**Comment by jerboaa** [comment]
`cpu_limit == null ? physical_cores : (cpu_limit / 1024)` => it does exactly that. setting CPU requests in OpenShift is mapped to a CPU limit in the container.

**Comment by jerboaa** [comment]
Use `-Xlog:os+container=trace` to trace what the JDK is doing. Not sure what your setup is and what you'd expect. It's not clear from the description.

**Comment by beikov** [comment]
So what is the CPU Limit in OpenShift mapped to then? I'd personally assume that the limit i.e. the upper bound is what should be use to determine the available processors. I tried to read through the mailing list discussion but I wasn't able to figure out exactly why mapping the cpu shares to available processors is a good idea.

When deploying a JVM to a normal host without cgroup limits/shares, you also get the number of physical cores as available processors returned, regardless of what priority the process has. So this decision seems counterintuitive to me.

**Comment by beikov** [comment]
Hey @karianna! Is there a chance I will get a decent answer as to why this is this way? It seems very counter intuitive to me that cpu-shares, which is AFAIU more or less a priority value for the linux scheduler, is used as a basis for available processors in the OpenJDK. IMO the value should be calculated by using `cpu_limit == null ? physical_cores : (cpu_limit / 1024)` just like @jerbaaa wrote.

I ruled out that this is _just_ OpenShift related as I was able to reproduce it with plain docker by setting the cpu-shares. Increasing cpu-shares or cpu-request in OpenShift to get the desired available processors being reported is not helpful, as that has an effect on scheduling in OpenShift. A 8 core node can only hold up pods with a sum of cpu requests of 8196. If some of the processes don't fully use their share(which they rarely will), the node will always be underutilized.
